File: frogoui/src/main/java/com/frogobox/ui/fireworks/modifiers/ScaleModifier.java
package com.frogobox.ui.fireworks.modifiers;

import android.view.animation.Interpolator;
import android.view.animation.LinearInterpolator;

import com.frogobox.ui.fireworks.Particle;

public class ScaleModifier implements ParticleModifier {

    private final float mInitialValue;
    private final float mFinalValue;
    private final long mEndTime;
    private final long mStartTime;
    private final long mDuration;
    private final float mValueIncrement;
    private final Interpolator mInterpolator;

    public ScaleModifier(float initialValue, float finalValue, long startMilis, long endMilis, Interpolator interpolator) {
        mInitialValue = initialValue;
        mFinalValue = finalValue;
        mStartTime = startMilis;
        mEndTime = endMilis;
        mDuration = mEndTime - mStartTime;
        mValueIncrement = mFinalValue - mInitialValue;
        mInterpolator = interpolator;
    }

    public ScaleModifier(float initialValue, float finalValue, long startMilis, long endMilis) {
        this(initialValue, finalValue, startMilis, endMilis, new LinearInterpolator());
    }

    @Override
    public void apply(Particle particle, long miliseconds) {
        if (miliseconds < mStartTime) {
            particle.mScale = mInitialValue;
        } else if (miliseconds > mEndTime) {
            particle.mScale = mFinalValue;
        } else {
            float interpolaterdValue = mInterpolator.getInterpolation((miliseconds - mStartTime) * 1f / mDuration);
            float newScale = mInitialValue + mValueIncrement * interpolaterdValue;
            particle.mScale = newScale;
        }
    }

}

File: frogoui/src/main/java/com/frogobox/ui/fireworks/modifiers/ScaleModifier.kt
package com.frogobox.ui.fireworks.modifiers

import android.view.animation.Interpolator
import android.view.animation.LinearInterpolator
import com.frogobox.ui.fireworks.Particle

class ScaleModifier @JvmOverloads constructor(
    private val mInitialValue: Float,
    private val mFinalValue: Float,
    private val mStartTime: Long,
    private val mEndTime: Long,
    interpolator: Interpolator = LinearInterpolator()
) : ParticleModifier {

    private val mDuration: Long = mEndTime - mStartTime
    private val mValueIncrement: Float = mFinalValue - mInitialValue
    private val mInterpolator: Interpolator

    init {
        mInterpolator = interpolator
    }

    override fun apply(particle: Particle?, miliseconds: Long) {
        if (miliseconds < mStartTime) {
            particle!!.mScale = mInitialValue
        } else if (miliseconds > mEndTime) {
            particle!!.mScale = mFinalValue
        } else {
            val interpolaterdValue =
                mInterpolator.getInterpolation((miliseconds - mStartTime) * 1f / mDuration)
            val newScale = mInitialValue + mValueIncrement * interpolaterdValue
            particle!!.mScale = newScale
        }
    }
}